In [ ]:
%load_ext autoreload
%autoreload 2

In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_svmlight_file
from sklearn.model_selection import train_test_split
import sys; sys.path.insert(0,'..')
from main import *

In [2]:
X, y = load_svmlight_file('../data/phishing/phishing.txt')
X = X.toarray()

# Convert labels from {0, 1} to {-1, 1}
y = y * 2 - 1

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.5, stratify=y, 
                                                    shuffle=True, random_state=1)

In [3]:
svm_arr, tree_arr = data_size_experiment(X_train, y_train, X_test, y_test,
                                         experiment_type='table', kernel='linear',
                                         random_state=1, verbose=True)

Progress:
PRELIMINARY
Training accuracy of rank SVM: 0.94
GENERATE W
num (-1, -1): 7
num (-1,  1): 16
num ( 1, -1): 10
num ( 1,  1): 17
Training accuracy of rank SVM: 0.90

LINEAR SVM SUBMODELS
Using 20 classifiers and sample proportion of 1
t	Train		Test
1	0.91		0.90
2	0.83		0.79
3	0.83		0.85
4	0.50		0.45
5	0.90		0.90
6	0.80		0.76
7	0.84		0.85
8	0.61		0.62
9	0.79		0.80
10	0.77		0.78
11	0.84		0.82
12	0.64		0.66
13	0.81		0.80
14	0.62		0.66
15	0.82		0.81
16	0.65		0.69
17	0.82		0.78
18	0.60		0.61
19	0.80		0.78
20	0.60		0.61
t = 1 was best iteration with accuracy 0.90

-----------------------------------------------
1 / 40 complete
-----------------------------------------------

LINEAR SVM SUBMODELS
Using 20 classifiers and sample proportion of 1
t	Train		Test
1	0.92		0.92
2	0.74		0.71
3	0.89		0.88
4	0.51		0.50
5	0.90		0.92
6	0.71		0.72
7	0.88		0.88
8	0.47		0.47
9	0.88		0.89
10	0.83		0.82
11	0.90		0.86
12	0.50		0.57
13	0.73		0.72
14	0.87		0.86
15	0.80		0.79
16	0.55		0.56
17	0.71		0.69
18	

2	0.74		0.69
3	0.86		0.88
4	0.58		0.55
5	0.85		0.89
6	0.67		0.63
7	0.85		0.87
8	0.68		0.65
9	0.87		0.88
10	0.61		0.59
11	0.87		0.89
12	0.66		0.64
13	0.87		0.88
14	0.65		0.66
15	0.86		0.86
16	0.69		0.68
17	0.79		0.80
18	0.65		0.64
19	0.87		0.86
20	0.72		0.71
t = 1 was best iteration with accuracy 0.90

-----------------------------------------------
17 / 40 complete
-----------------------------------------------

LINEAR SVM SUBMODELS
Using 20 classifiers and sample proportion of 1
t	Train		Test
1	0.91		0.91
2	0.73		0.69
3	0.84		0.87
4	0.55		0.57
5	0.89		0.90
6	0.80		0.79
7	0.89		0.89
8	0.57		0.56
9	0.85		0.84
10	0.85		0.83
11	0.85		0.86
12	0.58		0.57
13	0.77		0.78
14	0.79		0.78
15	0.83		0.85
16	0.72		0.67
17	0.70		0.70
18	0.72		0.73
19	0.74		0.76
20	0.70		0.68
t = 1 was best iteration with accuracy 0.91

-----------------------------------------------
18 / 40 complete
-----------------------------------------------

LINEAR SVM SUBMODELS
Using 20 classifiers and sample proportion of 1
t

5	0.92		0.90
6	0.91		0.90
7	0.91		0.90
8	0.91		0.90
9	0.91		0.90
10	0.91		0.90
11	0.91		0.90
12	0.91		0.90
13	0.91		0.90
14	0.92		0.90
15	0.91		0.90
16	0.91		0.90
17	0.91		0.90
18	0.92		0.90
19	0.91		0.90
20	0.91		0.90
t = 1 was best iteration with accuracy 0.91

-----------------------------------------------
33 / 40 complete
-----------------------------------------------

LINEAR SVM SUBMODELS
Using 20 classifiers and sample proportion of 1
t	Train		Test
1	0.92		0.90
2	0.92		0.90
3	0.91		0.90
4	0.92		0.90
5	0.92		0.90
6	0.92		0.90
7	0.92		0.90
8	0.92		0.90
9	0.92		0.90
10	0.92		0.90
11	0.92		0.90
12	0.92		0.90
13	0.92		0.90
14	0.92		0.90
15	0.92		0.90
16	0.92		0.90
17	0.92		0.90
18	0.92		0.90
19	0.92		0.90
20	0.92		0.90
t = 1 was best iteration with accuracy 0.90

-----------------------------------------------
34 / 40 complete
-----------------------------------------------

LINEAR SVM SUBMODELS
Using 20 classifiers and sample proportion of 1
t	Train		Test
1	0.92		0.91
2	0.92		0.91


In [4]:
table_results(svm_arr, tree_arr)

(Performance on test set)

50 comparisons
	acc	sderr
SVM: 	90.9	0.2
Tree: 	nan	nan

200 comparisons
	acc	sderr
SVM: 	90.5	0.2
Tree: 	nan	nan


    50 comparisons
        acc	sderr
    SVM: 	88.6	0.3
    Tree: 	nan	nan

    200 comparisons
        acc	sderr
    SVM: 	90.5	0.1
    Tree: 	nan	nan

# New

In [ ]:
def generate_preliminary_pairs(X_train, y_train, kernel='linear', random_seed=1):
    """ Used to generate preliminary pairs from all training data """
    
    clf = SVC(C=1.0, kernel=kernel)
    clf.fit(X_train, y_train)

    print("PRELIMINARY")
    print("Training accuracy of rank SVM: %.2f" % clf.score(X_train, y_train))

    rank = clf.decision_function(X_train)
    
    print("Sign of predicted rank matches true labels with accuracy: %.2f\n"
          % accuracy_score(y_train, np.sign(rank)))
    
    return rank

In [ ]:
# Preliminary function
# Use entire training set to get pairs

# kernel = 'linear'
# random_seed = 1

def generate_preliminary_pairs(X_train, y_train, kernel='linear', random_seed=1):
    """ Used to generate preliminary pairs from all training data """
    
    np.random.seed(random_seed)

    pos_idx = np.squeeze(np.argwhere(y_train == 1))
    neg_idx = np.squeeze(np.argwhere(y_train == -1))
    pos_idx = pos_idx[np.random.permutation(pos_idx.shape[0])]
    neg_idx = neg_idx[np.random.permutation(neg_idx.shape[0])]

    cut = min(pos_idx.size, neg_idx.size)

    rnd_c = np.random.choice([1, -1], [pos_idx.shape[0], 1])[:cut]
    X_pair = (X_train[pos_idx, :][:cut] - X_train[neg_idx, :][:cut]) * rnd_c
    y_pair = rnd_c.ravel()

    clf = SVC(C=1.0, kernel=kernel)
    clf.fit(X_pair, y_pair)

    print("PRELIMINARY")
    print("Training accuracy of rank SVM: %.2f" % clf.score(X_pair, y_pair))

    rank = clf.decision_function(X_train)
    
    print("Sign of predicted rank matches true labels with accuracy: %.2f\n"
          % accuracy_score(y_train, np.sign(rank)))
    
    return rank

In [ ]:
r = generate_preliminary_pairs(X_train, y_train)

In [ ]:
def generate_from_preliminary_W(X_train, y_train, rank, m, kernel='linear', random_seed=1):
    """ Modified version of 'generate_svm_rank_W' that uses preliminary pairs """
    
    np.random.seed(random_seed)
    
    pair1 = np.random.choice(X_train.shape[0], size=m)
    pair2 = np.random.choice(X_train.shape[0], size=m)

    X_train1 = X_train[pair1]
    X_train2 = X_train[pair2]
    
    # Print num of each type of pair
    y_stk = np.hstack((y_train[pair1].reshape(-1, 1), y_train[pair2].reshape(-1, 1)))
    y_unq, y_cts = np.unique(y_stk, return_counts=True, axis=0)
    print('GENERATE W')
    print('num (-1, -1): %d\nnum (-1,  1): %d\nnum ( 1, -1): %d\nnum ( 1,  1): %d'
          % (y_cts[0], y_cts[1], y_cts[2], y_cts[3]))

    X_diff = X_train1 - X_train2

    # If X1 ranked above X2, then y_diff is positive
    y_diff = np.sign(rank[pair1] - rank[pair2])

    clf = SVC(C=1.0, kernel=kernel)
    clf.fit(X_diff, y_diff)

    print("Training accuracy of rank SVM: %.2f\n" % clf.score(X_diff, y_diff))

    rank = clf.decision_function(X_train)
    sigma = 1
    W = gen_conf_matrix(rank, sigma)
    
    return W

In [ ]:
W = generate_from_preliminary_W(X_train, y_train, r, 200)

In [ ]:
W.shape, r.shape

# Old

In [ ]:
def generate_svm_rank_W(X_train, y_train, m, kernel='linear', random_seed=1):
    """ Generates pairwise comparison matrix W using Mohammad's method: learned linear SVM ranker trained on m pairs """

    np.random.seed(random_seed)

    pos_idx = np.squeeze(np.argwhere(y_train == 1))
    neg_idx = np.squeeze(np.argwhere(y_train == -1))
    pos_idx = pos_idx[np.random.permutation(pos_idx.shape[0])[:m]]
    neg_idx = neg_idx[np.random.permutation(neg_idx.shape[0])[:m]]

    cut = min(pos_idx.size, neg_idx.size)
    if cut != m:
        print("Cut at %d rather than expected %d" % (cut, m))

    rnd_c = np.random.choice([1, -1], [pos_idx.shape[0], 1])[:cut]
    X_pair = (X_train[pos_idx,:][:cut] - X_train[neg_idx,:][:cut]) * rnd_c
    y_pair = rnd_c

    clf = SVC(C=1.0, kernel=kernel)
    clf.fit(X_pair, y_pair)

    print("Training accuracy of rank SVM: %.2f\n" % clf.score(X_pair, y_pair))

    rank = clf.decision_function(X_train)
    sigma = 1
    W = gen_conf_matrix(rank, sigma)

    return W

In [ ]:
svm_arr, tree_arr = data_size_experiment(X_train, y_train, X_test, y_test,
                                         experiment_type='table', kernel='linear',
                                         random_state=1, verbose=True)

In [ ]:
table_results(svm_arr, tree_arr)